libs/@local/hashql/core/src/span/table.rs

```diff
@@ -0,0 +1,613 @@
+
+        let Some(element) = self.spans.get_mut(index) else {
+            return false;
+        };
+
+        func(
+            &mut element.span,
+            SpanAncestorsMut {
+                spans: &mut self.ancestors[element.ancestors.clone()],
+                mode: &mut element.mode,
+            },
+        );
+        true
+    }
+
+    fn get_entry(&self, span: SpanId) -> Option<&SpanEntry<S>> {
+        if span.source_id() != self.source_id {
+            return None;
+        }
+
+        let index = span.id() as usize;
+
+        self.spans.get(index)
+    }
+
+    /// Retrieves span data by ID.
+    ///
+    /// Returns a reference to the span data associated with the given [`SpanId`].
+    ///
+    /// # Returns
+    ///
+    /// - `Some(&S)` if the span exists and belongs to this table's source
+    /// - `None` if the span doesn't exist or belongs to a different source
+    ///
+    /// # Examples
+    ///
+    /// Successful lookup:
+    ///
+    /// ```rust
+    /// use hashql_core::span::{Span, SpanAncestors, SpanTable, TextRange};
+    /// use hashql_diagnostics::source::SourceId;
+    ///
+    /// # struct MySpan { value: i32, range: TextRange }
+    /// # impl hashql_core::span::Span for MySpan {
+    /// #     fn range(&self) -> TextRange { self.range }
+    /// # }
+    /// let mut table = SpanTable::new(SourceId::new_unchecked(0));
+    ///
+    /// let span = MySpan {
+    ///     value: 42,
+    ///     range: TextRange::new(10.into(), 15.into()),
+    /// };
+    /// let span_id = table.insert(span, SpanAncestors::empty());
+    ///
+    /// let retrieved = table.get(span_id).expect("span exists");
+    /// assert_eq!(retrieved.value, 42);
+    /// assert_eq!(u32::from(retrieved.range().len()), 5u32);
+    /// ```
+    ///
+    /// Cross-source lookup failure:
+    ///
+    /// ```rust
+    /// use hashql_core::span::{SpanAncestors, SpanTable, TextRange};
+    /// use hashql_diagnostics::source::SourceId;
+    ///
+    /// # struct MySpan { range: TextRange }
+    /// # impl hashql_core::span::Span for MySpan {
+    /// #     fn range(&self) -> TextRange { self.range }
+    /// # }
+    /// let mut table = SpanTable::new(SourceId::new_unchecked(0));
+    /// let span = MySpan {
+    ///     range: TextRange::new(0.into(), 10.into()),
+    /// };
+    /// let span_id = table.insert(span, SpanAncestors::empty());
+    ///
+    /// // Attempt lookup from wrong table
+    /// let table2: SpanTable<MySpan> = SpanTable::new(SourceId::new_unchecked(1));
+    /// assert!(table2.get(span_id).is_none());
+    /// ```
+    #[must_use]
+    pub fn get(&self, span: SpanId) -> Option<&S> {
+        self.get_entry(span).map(|entry| &entry.span)
+    }
+
+    fn absolute_impl(&self, span: SpanId, depth: usize) -> Option<SourceSpan>
+    where
+        S: Span,
+    {
+        assert!(
+            depth <= 32,
+            "Cannot resolve excessively deep span of {depth}, likely due to a circular dependency"
+        );
```

Comment:
Can we make the maximum depths a constant? Or pass it to this function? Or both?